source/exec.tex

```diff
@@ -4904,32 +4906,42 @@
 \begin{codeblock}
 namespace std::execution {
   template<@\libconcept{scope_token}@ Token, @\libconcept{sender}@ Sender>
-  struct @\exposid{associate-data}@ {                                       // \expos
-    using @\exposid{wrap-sender}@ =                                         // \expos
+  struct @\exposid{associate-data}@ {                                               // \expos
+    using @\exposid{wrap-sender}@ =                                                 // \expos
       remove_cvref_t<decltype(declval<Token&>().wrap(declval<Sender>()))>;
+    using @\exposid{assoc-t}@ = decltype(declval<Token&>().try_associate());        // \expos
+    using @\exposid{sender-ref}@ =                                                  // \expos
+      unique_ptr<@\exposid{wrap-sender}@, decltype([](auto* p) noexcept { destroy_at(p); })>;
 
     explicit @\exposid{associate-data}@(Token t, Sender&& s)
       : @\exposid{sndr}@(t.wrap(std::forward<Sender>(s))),
-        @\exposid{token}@(t) {
-      if (!@\exposid{token}@.try_associate())
-        @\exposid{sndr}@.reset();
+        @\exposid{assoc}@([&] {
+        @\exposid{sender-ref}@ guard{addressof(@\exposid{sndr}@)};
+        auto assoc = t.try_associate();
```

Comment:
Should we choose a different name for this local variable 'assoc' which doesn't shadow the data member _assoc_?